src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java

```diff
@@ -60,91 +64,119 @@ public static Map< Edgel, Double > getCosts(
 			final double angleCutoffDistance,
 			final double maxCenterDistance )
 	{
 		final int numPointsPerSample = 9;
+		if ( edgels.size() < numPointsPerSample )
+			throw new NoEllipsoidFoundException( "Not enough edgels to fit an ellipsoid." );
 
 		final Cost costFunction = new EdgelDistanceCost( outsideCutoffDistance, insideCutoffDistance, angleCutoffDistance );
 
 		final Random rand = new Random( System.currentTimeMillis() );
-		final ArrayList< Integer > indices = new ArrayList<>();
+		final TIntArrayList indices = new TIntArrayList();
 		final double[][] coordinates = new double[ numPointsPerSample ][ 3 ];
 
 		Ellipsoid bestEllipsoid = null;
 		double bestCost = Double.POSITIVE_INFINITY;
 		final double[] center = new double[ 3 ];
+		int candidates = 0;
 
 		for ( int sample = 0; sample < numSamples; ++sample )
 		{
-			try
-			{
-				indices.clear();
-				for ( int s = 0; s < numPointsPerSample; ++s )
-				{
-					int i = rand.nextInt( edgels.size() );
-					while ( indices.contains( i ) )
-						i = rand.nextInt( edgels.size() );
-					indices.add( i );
-					edgels.get( i ).localize( coordinates[ s ] );
-				}
-				final Ellipsoid ellipsoid = FitEllipsoid.yuryPetrov( coordinates );
-
-				// don't count degenerate samples
-				final double[] radii = ellipsoid.getRadii();
-				if ( Double.isNaN( radii[ 0 ] ) || Double.isNaN( radii[ 1 ] ) || Double.isNaN( radii[ 2 ] ) )
-				{
-					--sample;
-					continue;
-				}
-
-				ellipsoid.localize( center );
-				LinAlgHelpers.subtract( center, expectedCenter, center );
-				if ( LinAlgHelpers.length( center ) > maxCenterDistance )
-					continue;
-
-				final double cost = costFunction.compute( ellipsoid, edgels );
-				if ( cost < bestCost )
-				{
-					bestCost = cost;
-					bestEllipsoid = ellipsoid;
-				}
-			}
-			catch ( final IllegalArgumentException e )
-			{
-//				e.printStackTrace();
-//				System.out.println( "oops" );
-			}
-			catch ( final RuntimeException e )
+			sampleCoordinatesFromEdgels( edgels, coordinates, rand, indices );
+
+			final Ellipsoid ellipsoid = tryFitEllipsoidYuryPetrov( coordinates );
+
+			if ( ! isEllipsoidValid( ellipsoid, expectedCenter, maxCenterDistance, center ) )
+				continue;
+
+			final double cost = costFunction.compute( ellipsoid, edgels );
+			if ( cost < bestCost )
 			{
-//				System.out.println( "psd" );
+				bestCost = cost;
+				bestEllipsoid = ellipsoid;
 			}
+
+			if( ++candidates >= numCandidates )
+				break;
 		}
 
-		if ( bestEllipsoid == null )
-			return null;
+		if ( bestEllipsoid == null ) // no ellipsoid found
+			throw new NoEllipsoidFoundException( "No ellipsoid found, that is near to the expected center." );
 
-		final Ellipsoid refined = fitToInliers( bestEllipsoid, edgels, outsideCutoffDistance, insideCutoffDistance, angleCutoffDistance );
-		if ( refined == null )
+		try
+		{
+			// refined ellipsoid
+			return fitToInliers( bestEllipsoid, edgels, outsideCutoffDistance, insideCutoffDistance,
+					angleCutoffDistance );
```

Comment:
```suggestion
			// refined ellipsoid
			Ellipsoid refinedEllipsoid = fitToInliers( edgels, bestEllipsoid, costFunction );

			// test, if refined ellipsoid is valid
			if ( isEllipsoidValid( refinedEllipsoid, expectedCenter, maxCenterDistance, center ) )
				return refinedEllipsoid;

			// return best ellipsoid without refinement, if refined ellipsoid is not valid
			return bestEllipsoid;
```
While testing with a large dataset, I figure out that it makes sense to test, if the refined ellipsoid is valid, since it may cause problems in subsequent steps (e.g. viewing the ellipsoid in BDV), if it is not valid.

In case, it is not valid, there are two options: not returning an ellipsoid at all, or returning the best ellipsoid without refinement. It is hard to decide, which option is better without testing. I would recommend to return the best ellipsoid without refinement in this case. The testing which option is better could be a new issue in this repo.

Reply:
Good change. Why do you suggests to add all the comments, that say exactly whats written below?

Reply:
You are right. Some of them were really too obvious to be written. It is okay for me as it is now. Please resolve.